tests/test_cli.py

```diff
@@ -0,0 +1,114 @@
+        parser.assertCounts(1, 1)
+
+
+class TestClientArguments(unittest.TestCase):
+    """
+    Tests the client cli can parse all arguments it is supposed to
+    and can initialize the runner in preparation for a request
+    """
+    def setUp(self):
+        # initialize the client parser
+        self.parser = StubArgumentParser()
+        cli.client_main(self.parser)
+
+    def tearDown(self):
+        # include arguments common to all commands
+        inputStr = "--verbose --workarounds WORK,AROUND --key KEY {0} URL"
+        cliInput = inputStr.format(self.cliInput)
+        splits = cliInput.split()
+
+        # parse the arguments
+        args = self.parser.parser.parse_args(splits)
+
+        # invoke the initializer of the runner (which also parses args)
+        runner = args.runner(args)
+
+        # ensure the correct attributes on the runner are set
+        assert isinstance(runner._request, protocol.ProtocolElement)
+        assert isinstance(runner._httpClient, client.HTTPClient)
+
+    def testVariantsSearchArguments(self):
+        self.cliInput = """variants-search --referenceName REFERENCENAME
+        --variantName VARIANTNAME --callSetIds CALL,SET,IDS --start 0
+        --end 1 --pageSize 2"""
+
+    def testVariantSetsSearchArguments(self):
+        self.cliInput = """variantsets-search --pageSize 1 --datasetIds
+        DATA,SET,IDS"""
+
+    def testReferenceSetsSearchArguments(self):
+        self.cliInput = """referencesets-search --pageSize 1 --accessions
+        ACC,ESS,IONS --md5checksums MD5,CHECKSUMS --assemblyId ASSEMBLYID"""
+
+    def testReferencesSearchArguments(self):
+        self.cliInput = """references-search --pageSize 1 --accessions
+        ACC,ESS,IONS --md5checksums MD5,CHECKSUMS"""
+
+    def testReadGroupSetsSearchArguments(self):
+        self.cliInput = """readgroupsets-search --pageSize 1 --datasetIds
+        DATA,SET,IDS --name NAME"""
+
+    def testCallSetsSearchArguments(self):
+        self.cliInput = """callsets-search --pageSize 1 --name NAME
+        --variantSetIds VARIANT,SET,IDS"""
+
+    def testReadsSearchArguments(self):
+        self.cliInput = """reads-search --pageSize 1 --start 2 --end 3
+        --readGroupIds READ,GROUP,IDS --referenceId REFERENCEID
+        --referenceName REFERENCENAME"""
+
+
+class StubArgumentParser(object):
+    """
+    A stand-in object for an ArgumentParser that intercepts calls
+    to parse_args and print_help, but otherwise provides normal
+    behavior via passing through calls to an attribute ArgumentParser
+    """
+    def __init__(self):
+        self.parser = argparse.ArgumentParser(description="Stub")
+        self.helpCount = 0
+        self.parseArgsCount = 0
+        self.parse_args = self._parseArgs
+        self.print_help = self._help
+
+    def _help(self):
+        self.helpCount += 1
+
+    def _parseArgs(self):
+        self.parseArgsCount += 1
+        return ()
+
+    def assertCounts(self, parseArgsCount, helpCount):
```

Comment:
This class is an excellent idea! I wonder if it would be clearer to have two `assert` method here --- it's not clear reading the client code what exactly we're asserting. Perhaps `assertParseArgsCountEqual()` and `assertPrintHelpCountEqual()`? This will scale better if we want to add more assertions later.